### ISSUE TITLE
terraform crash with encrypted aws_ami_copy

### ISSUE THREAD

**Issue by rbuels** [hashicorp/terraform#7980]
Terraform 0.7.0 crashed for me just now while it was creating an `aws_ami_copy` with the `encrypted` and `kms_key_id` options set.

It crashed after exactly 5 minutes, the default connection timeout, which might be significant.

The console output was:
```
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (3m30s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (3m40s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (3m50s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m0s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m10s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m20s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m30s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m40s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (4m50s elapsed)
module.ecs_cluster.aws_ami_copy.ecs_ami: Still creating... (5m0s elapsed)
Error applying plan:

1 error(s) occurred:

* aws_ami_copy.ecs_ami: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
panic: runtime error: invalid memory address or nil pointer dereference
2016/08/04 11:35:24 [DEBUG] plugin: terraform: [signal 0xb code=0x1 addr=0x0 pc=0x370914]
2016/08/04 11:35:24 [DEBUG] plugin: terraform: 
2016/08/04 11:35:24 [DEBUG] plugin: terraform: goroutine 569 [running]:
2016/08/04 11:35:24 [DEBUG] plugin: terraform: panic(0x2fb1ba0, 0xc8200120f0)
2016/08/04 11:35:24 [DEBUG] plugin: terraform: 	/opt/go/src/runtime/panic.go:481 +0x3e6
2016/08/04 11:35:24 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsAmiRead(0xc8210c7bc0, 0x253e380, 0xc8207b7080, 0x0, 0x0)
```
(and then the debug output in the crash.log)

I put a tarball of the crash log and all of my terraform code on Dropbox, encrypted with the HashiCorp GPG key, for you guys to have a look at.  It's downloadable at https://www.dropbox.com/s/1u1t27zut2v35g4/terraform_crash.tar.gz.crypt?dl=0

Thanks in advance!

**Comment by rbuels** [comment]
I just ran it again, and got the same crash after 5m 10s, not exactly 5m, so the 5m thing might be a red herring, I don't know.

**Comment by kwilczynski** [comment]
@rbuels it might sound like a silly question to ask - but does it complete regardless of the Terraform crash? I wonder if the copy process is still pending on the side of AWS despite the crash.

**Comment by rbuels** [comment]
The ami copy did seem to complete in AWS, yes.

**Comment by rbuels** [comment]
Hmm, I can't start EC2 instances with the AMI that it produced, though.  Instances fail to start, with a generic "Server.InternalError: Internal error on launch"

**Comment by kwilczynski** [comment]
@rbuels oh bummer! Does it work with awscli, given the same set of parameters et al?

**Comment by rbuels** [comment]
hmm, when I try it with awscli it gives me:

```
$ AWS_PROFILE=default aws ec2 copy-image --encrypted --kms-key-id 1245-abcdef-fake --name ecs-encrypted --source-image-id ami-5d1b984a --source-region us-east-1

An error occurred (InvalidRequest) when calling the CopyImage operation: Images from AWS Marketplace cannot be copied to another AWS account.
```

**Comment by kwilczynski** [comment]
@rbuels if this is indeed a Market Place AMI and you want to use it directly, then you would need to first start an EC2 instance using this AMI and then take a snapshot to turn in into a private AMI, so that it becomes owned by you. As per [Copying an AMI Across AWS Accounts](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/CopyingAMIs.html#copy-ami-across-accounts).

If, on the contrary, it is your own AMI, then I would wager that there is a time out we don't handle well, or we get some response we do not handle. Perhaps there is also some other issue e.g. we send a request to AWS but the connection is interrupted/closed/times out, thus the EOF (which would be odd, as this is just HTTPS and we do not maintain a persistent connection to the backed, as far as I know, thus each request is a new connection).

I have tested `aws_ami_copy` just now (morning time in Tokyo) with my private AMI built with Packer, and I moved it from `ap-northeast-1` to `us-east-1` (which took a bit, due to latency, but it succeeded).

By the way, since I have no access to private GPG key that belongs to HashiCorp (since I do not work there), I cannot decrypt your file so that I could see the whole stack trace upon panic.

**Comment by rbuels** [comment]
I just tried it again using ami-55870742 as the source AMI, and the CLI appears to succeed with that, but terraform has a similar-looking crash.

**Comment by kwilczynski** [comment]
@rbuels got it! Let me try the same AMI, since for my own it's working fine.

Have you tried without the encryption and KSM? I wonder.

**Comment by rbuels** [comment]
Here's the most recent crash.log in the clear, it doesn't appear to have anything very sensitive in it.

[crash.log.txt](https://github.com/hashicorp/terraform/files/402857/crash.log.txt)

**Comment by rbuels** [comment]
Now I just tried it without the encryption and KMS, and got a similar crash.
[crash.log.txt](https://github.com/hashicorp/terraform/files/402863/crash.log.txt)

**Comment by kwilczynski** [comment]
@rbuels OK! Got a crash with the same AMI as you did. Around the 5 min mark.

Given the following template:

```
resource "aws_ami_copy" "test" {
    name = "terraform-test-123"
    description = "Test 123"
    source_ami_id = "ami-55870742"
    source_ami_region = "us-east-1"
    tags {
        Name = "Test 123"
    }
}
```

```
panic: runtime error: invalid memory address or nil pointer dereference
```

At the same place as in your case, for both 6.x and 7.x (built from master).

But... This is surprisingly erratic, because 1 in 3 attempts succeed for me.

**Comment by kwilczynski** [comment]
Firstly, apologies that you have experienced an issue!

Secondly, after looking into this we've found the root cause for the crash and the nil pointer being dereferenced.

The ECS image (AMI) you are using is a little bit special in terms how the block device mappings are presented via the API. By default, each EBS volume in the volume mappings, should include the snapshot ID element. In this case, the "/dev/xvdcz" volume (not the name "xvdcz" in itself, though) which is attached when the ECS instance is booted, does not have snapshot ID included, as per:

(as the Go SDK sees the response)

```
2016/08/05 10:04:16 [DEBUG] plugin: terraform:   DeviceName: "/dev/xvdcz",
2016/08/05 10:04:16 [DEBUG] plugin: terraform:   Ebs: {
2016/08/05 10:04:16 [DEBUG] plugin: terraform:     DeleteOnTermination: true,
2016/08/05 10:04:16 [DEBUG] plugin: terraform:     Encrypted: false,
2016/08/05 10:04:16 [DEBUG] plugin: terraform:     VolumeSize: 22,
2016/08/05 10:04:16 [DEBUG] plugin: terraform:     VolumeType: "gp2"
2016/08/05 10:04:16 [DEBUG] plugin: terraform:   }
2016/08/05 10:04:16 [DEBUG] plugin: terraform: }
```

(and the device mapping from the response as it is in XML)

```
2016/08/05 10:04:16 [DEBUG] plugin: terraform:             <blockDeviceMapping>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                 <item>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     <deviceName>/dev/xvda</deviceName>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     <ebs>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <snapshotId>snap-122abd42</snapshotId>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <volumeSize>8</volumeSize>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <deleteOnTermination>true</deleteOnTermination>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <volumeType>gp2</volumeType>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <encrypted>false</encrypted>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     </ebs>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                 </item>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                 <item>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     <deviceName>/dev/xvdcz</deviceName>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     <ebs>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <volumeSize>22</volumeSize>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <deleteOnTermination>true</deleteOnTermination>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <volumeType>gp2</volumeType>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                         <encrypted>false</encrypted>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                     </ebs>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:                 </item>
2016/08/05 10:04:16 [DEBUG] plugin: terraform:             </blockDeviceMapping>
```

The lack of snapshot ID causes the following to fail:

(from the stack trace)

`./src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_ami.go:178 +0x1293`

This translates to:

(see: [resource_aws_ami.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_ami.go))

```go
        for _, blockDev := range image.BlockDeviceMappings {
                if blockDev.Ebs != nil {
                        ebsBlockDev := map[string]interface{}{
                                "device_name":           *blockDev.DeviceName,
                                "delete_on_termination": *blockDev.Ebs.DeleteOnTermination,
                                "encrypted":             *blockDev.Ebs.Encrypted,
                                "iops":                  0,
                                "snapshot_id":           *blockDev.Ebs.SnapshotId,
                                "volume_size":           int(*blockDev.Ebs.VolumeSize),
                                "volume_type":           *blockDev.Ebs.VolumeType,
                        }
                        if blockDev.Ebs.Iops != nil {
                                ebsBlockDev["iops"] = int(*blockDev.Ebs.Iops)
                        }
                        ebsBlockDevs = append(ebsBlockDevs, ebsBlockDev)
                } else {
                        ephemeralBlockDevs = append(ephemeralBlockDevs, map[string]interface{}{
                                "device_name":  *blockDev.DeviceName,
                                "virtual_name": *blockDev.VirtualName,
                        })
                }
        }
```

And more precisely to this element of code:

`                                "snapshot_id":           *blockDev.Ebs.SnapshotId,`

This is where the nil pointer is being dereferenced. This is also why I could not reproduce this initially with my private and custom built images (non of them exhibit lack of the snapshot ID in the response).

I am going to open a PR that includes a fix for this.

Worth adding is that this "/dev/xvdcz" is very special to the ECS images, see: [Amazon ECS-optimized AMI](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html).

Packer users have stumbled upon this problem too (Google Groups): [amazon-ebs builder block device mapping with no snapshot id?](https://groups.google.com/forum/#!msg/packer-tool/2BkzHot5e9U/H5tbw850BAAJ).

Generally, I would recommend against copying these custom and dedicated ECS instances unless you absolutely must to do so. An alternative, would be to create an AMI that is ECS-enabled with Packer. These machines would be a normal EC2 instances running EC2 Agent software (see: [Amazon ECS Container Agent](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_agent.html)) and having some specialised configuration and tweaks done to them as per the AWS ECS requirements. Creating a custom AMI image might be something that might work better as users reported oddities with using this dedicated volume that AWS attaches to ECS instances. Although, there is some time and learning curve here, which might not be desirable.

Hope this helps!

Some more information for posterity:

- [Amazon ECS Container Instances](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_instances.html)
- [Amazon EC2 API Reference - DescribeImages](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeImages.html)

**Comment by kwilczynski** [comment]
@rbuels I am happy to say that with the fix in the PR #7995, I was able to copy this troublesome image successfully:

```
aws_ami_copy.test: Refreshing state... (ID: ami-9c7fb6fc)
aws_ami_copy.test: Creating...
  architecture:             "" => "<computed>"
  description:              "" => "Test 123"
  ebs_block_device.#:       "" => "<computed>"
  encrypted:                "" => "false"
  ephemeral_block_device.#: "" => "<computed>"
  image_location:           "" => "<computed>"
  kernel_id:                "" => "<computed>"
  kms_key_id:               "" => "<computed>"
  manage_ebs_snapshots:     "" => "<computed>"
  name:                     "" => "terraform-test-123"
  ramdisk_id:               "" => "<computed>"
  root_device_name:         "" => "<computed>"
  source_ami_id:            "" => "ami-55870742"
  source_ami_region:        "" => "us-east-1"
  sriov_net_support:        "" => "<computed>"
  tags.%:                   "" => "1"
  tags.Name:                "" => "Test 123"
  virtualization_type:      "" => "<computed>"
aws_ami_copy.test: Still creating... (10s elapsed)
aws_ami_copy.test: Still creating... (20s elapsed)
aws_ami_copy.test: Still creating... (30s elapsed)
aws_ami_copy.test: Still creating... (40s elapsed)
aws_ami_copy.test: Still creating... (50s elapsed)
aws_ami_copy.test: Still creating... (1m0s elapsed)
aws_ami_copy.test: Still creating... (1m10s elapsed)
aws_ami_copy.test: Still creating... (1m20s elapsed)
aws_ami_copy.test: Still creating... (1m30s elapsed)
aws_ami_copy.test: Still creating... (1m40s elapsed)
aws_ami_copy.test: Still creating... (1m50s elapsed)
aws_ami_copy.test: Still creating... (2m0s elapsed)
aws_ami_copy.test: Still creating... (2m10s elapsed)
aws_ami_copy.test: Still creating... (2m20s elapsed)
aws_ami_copy.test: Still creating... (2m30s elapsed)
aws_ami_copy.test: Still creating... (2m40s elapsed)
aws_ami_copy.test: Still creating... (2m50s elapsed)
aws_ami_copy.test: Still creating... (3m0s elapsed)
aws_ami_copy.test: Still creating... (3m10s elapsed)
aws_ami_copy.test: Still creating... (3m20s elapsed)
aws_ami_copy.test: Still creating... (3m30s elapsed)
aws_ami_copy.test: Still creating... (3m40s elapsed)
aws_ami_copy.test: Still creating... (3m50s elapsed)
aws_ami_copy.test: Still creating... (4m0s elapsed)
aws_ami_copy.test: Still creating... (4m10s elapsed)
aws_ami_copy.test: Still creating... (4m20s elapsed)
aws_ami_copy.test: Still creating... (4m30s elapsed)
aws_ami_copy.test: Still creating... (4m40s elapsed)
aws_ami_copy.test: Still creating... (4m50s elapsed)
aws_ami_copy.test: Still creating... (5m0s elapsed)
aws_ami_copy.test: Still creating... (5m10s elapsed)
aws_ami_copy.test: Still creating... (5m20s elapsed)
aws_ami_copy.test: Still creating... (5m30s elapsed)
aws_ami_copy.test: Still creating... (5m40s elapsed)
aws_ami_copy.test: Creation complete

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

**Comment by stack72** [comment]
Closed via #7995
